recipe/thirdparty/providers/oauth2_impl.go

```diff
@@ -0,0 +1,203 @@
+		if v == nil {
+			delete(queryParams, k)
+		} else {
+			queryParams[k] = v
+		}
+	}
+
+	url := config.AuthorizationEndpoint
+
+	/* Transformation needed for dev keys BEGIN */
+	if isUsingDevelopmentClientId(config.ClientID) {
+		queryParams["client_id"] = getActualClientIdFromDevelopmentClientId(config.ClientID)
+		queryParams["actual_redirect_uri"] = url
+		url = DevOauthAuthorisationUrl
+	}
+	/* Transformation needed for dev keys END */
+
+	queryParamsStr, err := qs.Marshal(queryParams)
+	if err != nil {
+		return tpmodels.TypeAuthorisationRedirect{}, err
+	}
+
+	return tpmodels.TypeAuthorisationRedirect{
+		URLWithQueryParams: url + "?" + queryParamsStr,
+		PKCECodeVerifier:   pkceCodeVerifier,
+	}, nil
+}
+
+func oauth2_ExchangeAuthCodeForOAuthTokens(config tpmodels.ProviderConfigForClient, redirectURIInfo tpmodels.TypeRedirectURIInfo, userContext supertokens.UserContext) (tpmodels.TypeOAuthTokens, error) {
+	tokenAPIURL := config.TokenEndpoint
+	accessTokenAPIParams := map[string]interface{}{
+		"client_id":    config.ClientID,
+		"redirect_uri": redirectURIInfo.RedirectURIOnProviderDashboard,
+		"code":         redirectURIInfo.RedirectURIQueryParams["code"].(string),
+		"grant_type":   "authorization_code",
+	}
+	if config.ClientSecret != "" {
+		accessTokenAPIParams["client_secret"] = config.ClientSecret
+	}
+	if redirectURIInfo.PKCECodeVerifier != nil {
+		accessTokenAPIParams["code_verifier"] = *redirectURIInfo.PKCECodeVerifier
+	}
+
+	for k, v := range config.TokenParams {
+		if v == nil {
+			delete(accessTokenAPIParams, k)
+		} else {
+			accessTokenAPIParams[k] = v
+		}
+	}
+
+	/* Transformation needed for dev keys BEGIN */
+	if isUsingDevelopmentClientId(config.ClientID) {
+		accessTokenAPIParams["client_id"] = getActualClientIdFromDevelopmentClientId(config.ClientID)
+		accessTokenAPIParams["redirect_uri"] = DevOauthRedirectUrl
+	}
+	/* Transformation needed for dev keys END */
+
+	oAuthTokens, err := doPostRequest(tokenAPIURL, accessTokenAPIParams, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return oAuthTokens, nil
+}
+
+func oauth2_GetUserInfo(config tpmodels.ProviderConfigForClient, oAuthTokens tpmodels.TypeOAuthTokens, userContext supertokens.UserContext) (tpmodels.TypeUserInfo, error) {
+	accessToken, accessTokenOk := oAuthTokens["access_token"].(string)
+	idToken, idTokenOk := oAuthTokens["id_token"].(string)
+
+	rawUserInfoFromProvider := tpmodels.TypeRawUserInfoFromProvider{}
+
+	if idTokenOk && config.JwksURI != "" {
+		claims := jwt.MapClaims{}
+		jwksURL := config.JwksURI
+		jwks, err := getJWKSFromURL(jwksURL)
+		if err != nil {
+			return tpmodels.TypeUserInfo{}, err
+		}
+		token, err := jwt.ParseWithClaims(idToken, claims, jwks.Keyfunc)
+		if err != nil {
+			return tpmodels.TypeUserInfo{}, err
+		}
+
+		if !token.Valid {
+			return tpmodels.TypeUserInfo{}, errors.New("invalid id_token supplied")
+		}
+		rawUserInfoFromProvider.FromIdTokenPayload = map[string]interface{}(claims)
```

Comment:
Should call the user's id token validation function here

Reply:
Done